Storage/FirebaseStorageUI/FUIStorageImageLoader.m

```diff
@@ -85,7 +86,8 @@ - (BOOL)canRequestImageForURL:(NSURL *)url {
     size = self.defaultMaxImageSize;
   }
   // Download the image from Firebase Storage
-  
+  // Each download task use independent serial coder queue, to ensure callback in order during prorgessive decoding
+  dispatch_queue_t coderQueue = dispatch_queue_create("com.google.firebaseui.storage.coderQueue", DISPATCH_QUEUE_SERIAL);
```

Comment:
👍 

Maybe name this `decoderQueue`, since it's where all the image decoding happens. If you want to cancel previous decoding operations when a new operation is enqueued, you can use `NSOperationQueue` instead of a dispatch queue.

Reply:
Using `NSOperationQueue` is a choice if we can cancel all previous progressive decoding when FIRStorageDownloadTask is cancelled. Need some extra check code for `FIRStorageDownloadTask`'s status.

Is there public API for this ? I found one `FIRStorageDownloadTask.state` (in `FIRStorageTask_Private.h` 😓 ), which is a enum. Need we check the state and earily return in the operation block body ?

Reply:
I've done using NSOperationQueue, without checking the `FIRStorageDownloadTask.state`.  The reason:

1. Change the behavior, if we've already have decoding process, we don't decode current partial images. Means the there are always 0 or 1 task been executed, no queue to wait. Directlly ignore the dummy decoding. (Because user only need one picture at the time, right ?)

2. When `FIRStorageDownloadTask` get cancelled, they will no longer report progress callback, so we don't need extra check inside NSBlockOperation, just let it done.

This design is the same as SDWebImage itself, see: https://github.com/SDWebImage/SDWebImage/pull/2483

Reply:
![image](https://user-images.githubusercontent.com/6919743/76378651-f24c2500-6388-11ea-966a-47800be4c9a3.png)

Reply:
Thanks for looking into this!